crates/payload/basic/src/lib.rs

```diff
@@ -0,0 +1,412 @@
+    initialized_cfg: CfgEnv,
+    /// The parent block.
+    parent_block: Arc<Block>,
+    /// Block extra data.
+    extra_data: Bytes,
+    /// Requested attributes for the payload.
+    attributes: PayloadBuilderAttributes,
+    /// The chain spec.
+    chain_spec: Arc<ChainSpec>,
+}
+
+/// Builds the payload and sends the result to the given channel.
+fn build_payload<Pool, Client>(
+    client: Client,
+    pool: Pool,
+    config: PayloadConfig,
+    cancel: Cancelled,
+    to_job: oneshot::Sender<Result<BuiltPayload, PayloadBuilderError>>,
+) where
+    Client: StateProviderFactory,
+    Pool: TransactionPool,
+{
+    #[inline(always)]
+    fn try_build<Pool, Client>(
+        client: Client,
+        pool: Pool,
+        config: PayloadConfig,
+        cancel: Cancelled,
+    ) -> Result<BuiltPayload, PayloadBuilderError>
+    where
+        Client: StateProviderFactory,
+        Pool: TransactionPool,
+    {
+        let PayloadConfig {
+            initialized_block_env,
+            initialized_cfg,
+            parent_block: _,
+            extra_data: _,
+            attributes,
+            chain_spec,
+        } = config;
+
+        // TODO this needs to access the _pending_ state of the parent block hash
+        let state = client.latest()?;
+
+        let mut db = SubState::new(State::new(state));
+        let _post_state = PostState::default();
+
+        let mut cumulative_gas_used = 0;
+        let block_gas_limit: u64 = initialized_block_env.gas_limit.try_into().unwrap_or(u64::MAX);
+
+        let mut executed_txs = Vec::new();
+        let best_txs = pool.best_transactions();
+
+        for tx in best_txs {
+            // ensure we still have capacity for this transaction
+            if cumulative_gas_used + tx.gas_limit() > block_gas_limit {
+                // TODO: try find transactions that can fit into the block
+                break
+            }
+
+            // check if the job was cancelled, if so we can exit early
+            if cancel.is_cancelled() {
+                return Err(PayloadBuilderError::BuildJobCancelled)
+            }
+
+            // convert tx to a signed transaction
+            let tx = tx.to_recovered_transaction();
+            let env = Env {
+                cfg: initialized_cfg.clone(),
+                block: initialized_block_env.clone(),
+                tx: tx_env_with_recovered(&tx),
+            };
+
+            let mut evm = revm::EVM::with_env(env);
+            evm.database(&mut db);
+
+            // TODO skip invalid transactions
+            let res = evm.transact().map_err(PayloadBuilderError::EvmExecutionError)?;
+
+            // append transaction to the list of executed transactions
+            executed_txs.push(tx);
+
+            // append gas used
+            cumulative_gas_used += res.result.gas_used();
+        }
+
+        // Process withdrawals
+        if chain_spec.fork(Hardfork::Shanghai).active_at_timestamp(attributes.timestamp) {
```

Comment:
`initialized_cfg` already contains revm spec that can be used to check if it's shanghai or not. this allows making `PayloadConfig` slimmer

Reply:
TIL